src/storage/cache_proxy/cluster_manager.rs

```diff
@@ -0,0 +1,1096 @@
+        // If watch_master failed, we will retry to watch master
+        self.inner.watch_master().await?;
+
+        Ok(())
+    }
+}
+
+/// Inner struct of cluster manager
+#[derive(Debug)]
+#[allow(dead_code)]
+pub struct ClusterManagerInner {
+    /// Etcd client
+    kv_engine: Arc<KVEngineType>,
+    /// Node session, try to keep the session alive, we will set the session in run method
+    node_session: ArcSwapOption<Session>,
+    /// current node info
+    node: ArcSwap<Node>,
+    /// current node list, we will set the session in run method
+    node_list: ArcSwap<Vec<Node>>,
+    /// current hash ring, we will set the session in run method
+    hash_ring: ArcSwap<Ring<Node>>,
+    // TODO: Support distribute cluster config here
+}
+
+#[allow(dead_code)]
+impl ClusterManagerInner {
+    /// Create a new cluster manager
+    pub fn new(kv_engine: Arc<KVEngineType>, node: Node) -> Self {
+        let node_session = ArcSwapOption::empty();
+        let node = ArcSwap::<Node>::new(Arc::new(node));
+        let node_list = ArcSwap::<Vec<Node>>::new(Arc::new(Vec::new()));
+        let hash_ring = ArcSwap::<Ring<Node>>::new(Arc::new(Ring::default()));
+
+        Self {
+            kv_engine,
+            node_session,
+            node,
+            node_list,
+            hash_ring,
+        }
+    }
+
+    /// Prepare for the session and other data
+    pub async fn prepare(&self) -> DatenLordResult<()> {
+        // Prepare session
+        let session = self
+            .kv_engine
+            .create_session(SESSION_TIMEOUT_SEC)
+            .await
+            .with_context(|| format!("{} failed to create session", self.node.load().ip()))?;
+        self.node_session.store(Some(session));
+        Ok(())
+    }
+
+    /// Update node info
+    pub async fn update_node_in_cluster(&self) -> DatenLordResult<()> {
+        let node = self.node.load();
+        let key = &KeyType::CacheNode(node.ip().to_owned());
+        let node_session = self.node_session.load();
+        let Some(current_session) = node_session.as_ref().cloned() else {
+            let current_node = self.node.load();
+            warn!(
+                "Current {} session is invalid, return to endpoint",
+                current_node.ip()
+            );
+            return Err(DatenLordError::CacheClusterErr {
+                context: vec![format!("Current {} session is invalid", current_node.ip())],
+            });
+        };
+
+        self.kv_engine
+            .set(
+                key,
+                &ValueType::Json(serde_json::to_value(node.as_ref())?),
+                Some(SetOption {
+                    session: Some(current_session),
+                    prev_kv: false,
+                }),
+            )
+            .await
+            .with_context(|| "Failed to register node to etcd".to_owned())?;
+
+        Ok(())
+    }
+
+    /// Register current node to etcd and keep alive
+    pub async fn register(&self) -> DatenLordResult<()> {
```

Comment:
the loginc of register() and update_node_in_cluster() is exactly the same ?

Reply:
This was a legacy issue that has now been fixed.